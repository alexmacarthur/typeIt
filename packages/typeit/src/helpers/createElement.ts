import { Element } from "../types";

export default (el): Element => document.createElement(el);
