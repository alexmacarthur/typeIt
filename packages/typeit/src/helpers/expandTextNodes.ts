import createTextNode from "./createTextNode";
import { El } from "../types";

const expandTextNodes = (
  element: El,
  stringSpliterator: (str: string) => string[]
): El => {
  [...element.childNodes].forEach((child) => {
    if (child.nodeValue) {
      stringSpliterator(child.nodeValue).forEach((c) => {
        child.parentNode.insertBefore(createTextNode(c), child);
      });

      child.remove();
      return;
    }

    expandTextNodes(child as El, stringSpliterator);
  });

  return element;
};

export default expandTextNodes;
