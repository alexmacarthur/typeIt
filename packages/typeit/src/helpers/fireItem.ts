import { CURSOR_ANIMATION_RESTART_DELAY } from "../constants";
import { El, QueueItem, QueueMapPair } from "../types";
import beforePaint from "./beforePaint";
import rebuildCursorAnimation from "./rebuildCursorAnimation";

let execute = (queueItem: QueueItem) => queueItem.func?.call(this);

interface FireItemArgs {
  index: number;
  queueItems: QueueMapPair[];
  wait: (...args: any) => Promise<void>;
  cursor: El | undefined;
}

let fireItem = async ({
  index,
  queueItems,
  wait,
  cursor,
}: FireItemArgs): Promise<number> => {
  let queueItem = queueItems[index][1];
  let instantQueue = [];
  let tempIndex = index;
  let futureItem = queueItem;
  let shouldBeGrouped = () => futureItem && !futureItem.delay;

  // Crawl through the queue and group together all items that
  // do not have have a delay and can be executed instantly.
  while (shouldBeGrouped()) {
    instantQueue.push(futureItem);
    
    shouldBeGrouped() && tempIndex++;
    futureItem = queueItems[tempIndex] ? queueItems[tempIndex][1] : null;
  }
  
  if (instantQueue.length) {
    // All are executed together before the browser has a chance to repaint.
    await beforePaint(async () => {
      for (let q of instantQueue) {
        await execute(q);
      }
    });
    
    // Important! Because we moved into the future, the index
    // needs to be modified and returned for accurate remaining execution.
    return tempIndex - 1;
  }

  let fire = async (): Promise<{animation: Animation, timingOptions: object, frames: AnimationKeyFrame[]}> => {
    // An animation is only registered on the cursor when it's made visible. 
    // If the cursor has been disabled, there won't be one here.
    let animation = cursor?.getAnimations()[0];
    let timingOptions: object, frames: AnimationKeyFrame[];

    if(animation) {
      timingOptions = cursor ? {
        ...animation.effect.getComputedTiming(),
        delay: queueItem.shouldPauseCursor() ? CURSOR_ANIMATION_RESTART_DELAY : 0
      } : {};
      frames = cursor ? animation.effect.getKeyframes() : [];
    }
    
    await wait(async () => {
      // Pause the cursor while stuff is happening.
      if(queueItem.shouldPauseCursor()) {
        animation?.cancel();
      }

      await beforePaint(() => {    
        execute(queueItem);
      })
    }, queueItem.delay);

    return { animation, frames, timingOptions };
  }

  let { animation, frames, timingOptions } = await fire();

  animation && rebuildCursorAnimation({
    cursor,
    frames,
    timingOptions
  });

  return index;
};

export default fireItem;
