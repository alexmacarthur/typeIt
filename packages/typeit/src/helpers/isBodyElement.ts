export default (node): boolean => node.tagName === "BODY";
