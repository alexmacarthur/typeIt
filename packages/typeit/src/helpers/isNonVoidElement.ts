const isNonVoidElement = (el) => /<(.+)>(.*?)<\/(.+)>/.test(el.outerHTML);

export default isNonVoidElement;