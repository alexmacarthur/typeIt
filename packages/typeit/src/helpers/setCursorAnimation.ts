import { El } from "../types";

const DEFAULT_TIMING_OPTIONS: Partial<AnimationEffectTiming> = {
  iterations: Infinity,
  easing: "steps(2, start)",
  fill: "forwards",
};

const DEFAULT_FRAMES: AnimationKeyFrame[] = [
  { opacity: 0 },
  { opacity: 0 },
  { opacity: 1 },
];

/**
 * Create and return an animation for the cursor.
 */
let setCursorAnimation = ({
  cursor,
  frames = null,
  timingOptions = {},
}: {
  cursor: El;
  frames?: AnimationKeyFrame[] | null;
  timingOptions: Partial<AnimationEffectTiming>;
}): Animation | null => {
  let animation = cursor.animate(frames || DEFAULT_FRAMES, {
    ...DEFAULT_TIMING_OPTIONS,
    ...timingOptions,
  });

  animation.id = cursor.dataset.tiAnimationId;

  return animation;
};

export default setCursorAnimation;
