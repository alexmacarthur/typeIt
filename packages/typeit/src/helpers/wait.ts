/**
 * Fire a callback after a delay, and add the timeout ID to a referenced array.
 */
let wait = async (callback: Function, delay: number | undefined, timeouts) => {
  return new Promise<void>((resolve) => {
    let cb = async () => {
      await callback();

      resolve();
    };

    timeouts.push(setTimeout(cb, delay || 0) as unknown as number);
  });
};

export default wait;
