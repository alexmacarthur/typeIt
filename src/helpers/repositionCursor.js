export default (element, allChars, cursor, cursorPosition) => {
  if (!cursor) {
    return;
  }

  let characterIndex = cursorPosition;

  // Make sure we stop when we're at the beginning of the string.
  characterIndex =
    characterIndex > allChars.length ? allChars.length : characterIndex;

  let nodeToInsertBefore = allChars[characterIndex - 1];

  element.insertBefore(cursor, nodeToInsertBefore || null);
};
