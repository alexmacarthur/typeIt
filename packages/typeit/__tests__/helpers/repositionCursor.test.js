import repositionCursor from "../../src/helpers/repositionCursor";
import { walkElementNodes } from "../../src/helpers/chunkStrings";
import expandTextNodes from "../../src/helpers/expandTextNodes";
import { defaultExpandTextNodes, emojiFiveDigitString } from "./util";
import GraphemeSplitter from "grapheme-splitter";

let element, allCharacters;

describe("cursor movement default charset", () => {
  beforeEach(() => {
    setHTML`<span id="el">12345<i class="ti-cursor">|</i></span>`;

    element = defaultExpandTextNodes(document.querySelector("#el"));

    allCharacters = walkElementNodes(element, true);
  });

  it("Does not move cursor when stepsToMove is zero.", () => {
    repositionCursor(element, allCharacters, 0);

    let expected = `<span id="el">12345<i class="ti-cursor">|</i></span>`;
    expect(document.body.innerHTML).toEqual(expected);
  });

  it("Moves cursor three steps back.", () => {
    repositionCursor(element, allCharacters, 3);

    let expected = `<span id="el">12<i class="ti-cursor">|</i>345</span>`;

    expect(document.body.innerHTML).toEqual(expected);
  });

  it("Moves cursor three back and two forward.", () => {
    repositionCursor(element, allCharacters, 3);

    repositionCursor(element, allCharacters, 1);

    let expected = `<span id="el">1234<i class="ti-cursor">|</i>5</span>`;
    expect(document.body.innerHTML).toEqual(expected);
  });

  it("Stops moving when at end of string.", () => {
    repositionCursor(element, allCharacters, -100);

    let expected = `<span id="el">12345<i class="ti-cursor">|</i></span>`;
    expect(document.body.innerHTML).toEqual(expected);
  });
});

describe("cursor movement emoji", () => {
  beforeEach(() => {
    // Set the HTML raw to avoid pre-processing of the emoji.
    document.body.innerHTML = `<span id="el">${emojiFiveDigitString}<i class="ti-cursor">|</i></span>`;
    const splitter = new GraphemeSplitter();
    element = expandTextNodes(document.querySelector("#el"), (str) =>
      splitter.splitGraphemes(str)
    );

    allCharacters = walkElementNodes(element, true);
  });
  it("Does not move cursor when stepsToMove is zero.", () => {
    repositionCursor(element, allCharacters, 0);

    let expected = `<span id="el">1️⃣2️⃣3️⃣4️⃣5️⃣<i class="ti-cursor">|</i></span>`;
    expect(document.body.innerHTML).toEqual(expected);
  });

  it("Moves cursor three steps back.", () => {
    repositionCursor(element, allCharacters, 3);

    let expected = `<span id="el">1️⃣2️⃣<i class="ti-cursor">|</i>3️⃣4️⃣5️⃣</span>`;

    expect(document.body.innerHTML).toEqual(expected);
  });

  it("Moves cursor three back and two forward.", () => {
    repositionCursor(element, allCharacters, 3);

    repositionCursor(element, allCharacters, 1);

    let expected = `<span id="el">1️⃣2️⃣3️⃣4️⃣<i class="ti-cursor">|</i>5️⃣</span>`;
    expect(document.body.innerHTML).toEqual(expected);
  });

  it("Stops moving when at end of string.", () => {
    repositionCursor(element, allCharacters, -100);

    let expected = `<span id="el">${emojiFiveDigitString}<i class="ti-cursor">|</i></span>`;
    expect(document.body.innerHTML).toEqual(expected);
  });
});
