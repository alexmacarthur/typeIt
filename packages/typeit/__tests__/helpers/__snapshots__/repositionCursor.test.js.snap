// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Does not move cursor when stepsToMove is zero. 1`] = `"<span id=\\"el\\">12345<i class=\\"ti-cursor\\">|</i></span>"`;

exports[`Moves cursor three back and two forward. 1`] = `"<span id=\\"el\\">1234<i class=\\"ti-cursor\\">|</i>5</span>"`;

exports[`Moves cursor three steps back. 1`] = `"<span id=\\"el\\">12<i class=\\"ti-cursor\\">|</i>345</span>"`;

exports[`Stops moving when at end of string. 1`] = `"<span id=\\"el\\">12345<i class=\\"ti-cursor\\">|</i></span>"`;
